plugins/aws/sts_provisioner.go

```diff
@@ -2,71 +2,89 @@ package aws
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"os"
+	"time"
 
 	"github.com/1Password/shell-plugins/sdk"
 	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	confighelpers "github.com/99designs/aws-vault/v7/vault"
 	"github.com/aws/aws-sdk-go-v2/aws"
-	"github.com/aws/aws-sdk-go-v2/credentials"
 	"github.com/aws/aws-sdk-go-v2/service/sts"
-	"github.com/aws/aws-sdk-go-v2/service/sts/types"
 )
 
+const defaultProfileName = "default"
+
 type STSProvisioner struct {
-	TOTPCode  string
-	MFASerial string
+	profileName        string
+	newProviderFactory func(cacheState sdk.CacheState, cacheOps sdk.CacheOperations, fields map[sdk.FieldName]string) STSProviderFactory
 }
 
-func (p STSProvisioner) Provision(ctx context.Context, in sdk.ProvisionInput, out *sdk.ProvisionOutput) {
-	var cached types.Credentials
-	if ok := in.Cache.Get("sts", &cached); ok {
-		out.AddEnvVar("AWS_ACCESS_KEY_ID", *cached.AccessKeyId)
-		out.AddEnvVar("AWS_SECRET_ACCESS_KEY", *cached.SecretAccessKey)
-		out.AddEnvVar("AWS_SESSION_TOKEN", *cached.SessionToken)
+func NewSTSProvisioner(profileName string) STSProvisioner {
+	return STSProvisioner{
+		profileName: profileName,
+		newProviderFactory: func(cacheState sdk.CacheState, cacheOps sdk.CacheOperations, fields map[sdk.FieldName]string) STSProviderFactory {
```

Comment:
Do I interpret this name correctly that we have a factory for a factory?

My current understanding based on the names:
- A providerfactory returns a provider.
- The provider in turn provides a credential

One point of confusion is that the return type of this function is also a provider factory. Based on the fieldname, I'd expect the return type to be a provider, as it's what the providerfactory (the func) returns.

Reply:
Yes basically a `newProviderFactory` is a function that returns a provider factory. I thought the `new` in the beginning of the name would signal that it's a function and not a provider factory. A factory of a  provider factory, as you noted, if we choose to extrapolate. This is a better alternative to the `SetState` pattern, because the specific provider factory used by the `STSProvisioner` is a `CacheProviderFactory` and needs some initialization with `CacheState`, `CacheOperations` and `ItemFields`.

Do you think we could come up with better naming than `newProviderFactory`? The only other name I can come up with is `newProviderFactoryFunc` which doesn't seem to add a lot of value 🤔